pykern/pkunit.py

```diff
@@ -511,6 +391,144 @@ def work_dir():
+                pkio.mkdir_parent_only(self.actual_path)
+                self.actual = pykern.pkjson.dump_pretty(self.actual, filename=self.actual_path)
+            return True
+        return False
+
+    def _set_jinja(self):
+        if self.is_jinja:
+            import pykern.pkjinja
+
+            self.expect = pykern.pkjinja.render_file(self.expect_path, self.kwargs['j2_ctx'], strict_undefined=True)
+            if self.a:
+                pkio.write_text(self.actual_path, self.actual)
+            return True
+        return False
+
+    def _set_default(self):
+        self.expect = pkio.read_text(self.expect_path)
+        if self.a:
+            pkio.write_text(self.actual_path, self.actual)
+
+    def _convert_xlsx(self):
+        if self.expect_path.ext == '.csv' and not self.actual_path.check(file=True):
+            return self._xlsx_to_csv()
+        return False
+
+    def _compare(self):
+        if self.expect == self.actual:
+            return
+        c = f"diff '{self.expect_path}' '{self.actual_path}'"
+        x = self._get_message()
+        with os.popen(c) as f:
+            pkfail(
+                '{}',
+                f'''expect != actual:
+    {c}
+    {f.read()}
+    {x}
+    '''
+            )
+
+    def _get_message(self):
+        if self.is_jinja:
+            return '''
+    Implementation restriction: The jinja values are not filled in the diff
+    so the actual can't be copied to the expected to fix things.
+    '''
+        else:
+            return f'''
+    to update test data:
+        cp '{self.actual_path}' '{self.expect_path}'
+    '''
+
+    def _xlsx_to_csv_convert(self, actual_xlsx, sheet):
+        try:
+            import pandas
+        except ModuleNotFoundError:
+            pkfail('optional module=pandas must be installed to compare xlsx={}', actual_xlsx)
+
+        p = pandas.read_excel(
+            actual_xlsx,
+            index_col=None,
+            sheet_name=sheet,
+        )
+        p.columns = p.columns.map(lambda c: '' if 'Unnamed' in str(c) else str(c))
+        p.to_csv(
+            str(self.actual_path),
+            encoding='utf-8',
+            index=False,
+            line_terminator='\r\n',
+        )
+
+    def _xlsx_to_csv(self):
+        from pykern.pkdebug import pkdlog
+
+        try:
+            b = self.actual_path.new(ext='.xlsx')
+            m = _CSV_SHEET_ID.search(b.purebasename)
+            s = 0
+            if m:
+                b = b.new(purebasename=m.group(1))
+                s = int(m.group(2))
+            if b.check(file=True):
+                self._xlsx_to_csv_convert(b, s)
+                return True
+            return False
+        except Exception:
+            pkdlog('ERROR converting xlsx to csv expect={} actual={}', self.expect_path, self.actual_path)
+            raise
+
+    def _validate_and_set_paths(self, expect_path, *args, **kwargs):
+        self.kwargs = kwargs
+        self.a = 'actual' in self.kwargs
```

Comment:
self._have_actual_kwarg. Don't use single letter variables for attributes, keys.